py4DSTEM/process/preprocess/preprocess.py

```diff
@@ -149,5 +150,114 @@ def bin_data_real(datacube, bin_factor):
+
+        Returns:
+            datacube                     datacube              
+            mask (bool):                 (optional) the bad pixel mask
+    """
+
+    # Mean image over all probe positions
+    diff_mean = np.mean(datacube.data,axis=(0,1))
+
+    # Local median
+    diff_local_med = np.sort(np.vstack([
+        np.roll(diff_mean,(-1,-1),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 0,-1),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 1,-1),axis=(0,1)).ravel(),
+        np.roll(diff_mean,(-1, 0),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 0, 0),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 1, 0),axis=(0,1)).ravel(),
+        np.roll(diff_mean,(-1, 1),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 0, 1),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 1, 1),axis=(0,1)).ravel(),
+        \
+        np.roll(diff_mean,(-1,-2),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 0,-2),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 1,-2),axis=(0,1)).ravel(),
+        np.roll(diff_mean,(-1, 2),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 0, 2),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 1, 2),axis=(0,1)).ravel(),
+        \
+        np.roll(diff_mean,(-2,-1),axis=(0,1)).ravel(),
+        np.roll(diff_mean,(-2, 0),axis=(0,1)).ravel(),
+        np.roll(diff_mean,(-2, 1),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 2,-1),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 2, 0),axis=(0,1)).ravel(),
+        np.roll(diff_mean,( 2, 1),axis=(0,1)).ravel(),        
+        ]), axis=0)
+    diff_compare = np.reshape(diff_local_med[-ind_compare-1,:], diff_mean.shape)
+
+    # Generate mask
+    mask = diff_mean - diff_compare > thresh
+
+    # apply filtering
+    for ax, ay in tqdmnd(*(datacube.R_Nx,datacube.R_Ny), desc="Cleaning pixels", unit=" images"):
+        # Calculate local 3x3 median images
+        im_med = median_filter(
+            datacube.data[ax,ay,:,:],
+            size=3,
+            mode='nearest')
+        datacube.data[ax,ay,:,:][mask] = im_med[mask]
+
+    if return_mask is True:
+        return datacube, mask
+    else:
+        return datacube
+
+def datacube_diffraction_shift(
+    datacube,
+    xshifts,
+    yshifts,
+    periodic=True,
+    bilinear=False,
+    ):
+    """
+    This function performs pixel filtering to remove hot or cold pixels. We first compute a moving median filter,
+    applied to the mean diffraction image. Pixels with intensity int_0 are compared to a locally ordered median
+    value, default is the 2nd brightest pixel.
+
+    Args:
+            datacube (DataCube):      
+            xshifts (float):       array or scalar value for the x dim shifts
+            yshifts (float):       array or scalar value for the y dim shifts
+            periodic (bool):       flag for periodic boundary conditions
+            bilinear (bool):       flag for bilinear image shifts
```

Comment:
Maybe note that if bilinear is off, Fourier is used

Reply:
done